src/throttler/remote_throttler.js

```diff
@@ -0,0 +1,179 @@
+import Utils from '../util.js';
+
+const DEFAULT_REFRESH_INTERVAL_MS = 5000;
+const DEFAULT_INITIAL_DELAY_MS = 5000;
+const DEFAULT_THROTTLER_HOST = '0.0.0.0';
+const DEFAULT_THROTTLER_PORT = 5778;
+
+// UNIT_CREDIT is the minimum amount of credits necessary to not be throttled.
+// i.e. if currentCredits > UNIT_CREDIT, then the operation will not be throttled.
+const UNIT_CREDIT = 1.0;
+
+type CreditsPerOperation = { [key: string]: number };
+
+export default class RemoteThrottler {
+  _serviceName: string;
+  _logger: Logger;
+  _metrics: Metrics;
+
+  _refreshIntervalMs: number;
+  _host: string;
+  _port: number;
+
+  _uuid: string;
+  _credits: CreditsPerOperation;
+
+  _initialDelayTimeoutHandle: any;
+  _refreshIntervalHandle: any;
+  _onCreditsUpdate: ?Function;
+
+  /**
+   * Creates a RemoteThrottler that fetches credits remotely from jaeger-agent.
+   *
+   * @param {string} [serviceName] - name of the current service / application, same as given to Tracer
+   * @param {object} [options] - optional settings
+   * @param {object} [options.logger] - optional logger, see _flow/logger.js
+   * @param {object} [options.metrics] - instance of Metrics object
+   * @param {number} [options.refreshIntervalMs] - interval in milliseconds that determines how often credits
+   * are fetched
+   * @param {number} [options.initialDelayMs] - interval in milliseconds that determines how soon after initialization
+   * credits are first fetched
+   * @param {string} [options.host] - host for jaeger-agent, defaults to 'localhost'
+   * @param {number} [options.port] - port for jaeger-agent for /credits endpoint
+   * @param {function} [options.onCreditsUpdate] - callback function once credits are updated. Used for testing.
+   */
+  constructor(serviceName: string, options: any = {}) {
+    this._serviceName = serviceName;
+    this._logger = options.logger || new NullLogger();
+    this._metrics = options.metrics || new Metrics(new NoopMetricFactory());
+    this._refreshIntervalMs = options.refreshIntervalMs || DEFAULT_REFRESH_INTERVAL_MS;
+    this._host = options.host || DEFAULT_THROTTLER_HOST;
+    this._port = options.port || DEFAULT_THROTTLER_PORT;
+
+    this._credits = {};
+    this._onCreditsUpdate = options.onCreditsUpdate;
+
+    this._initialDelayTimeoutHandle = setTimeout(
+      this._afterInitialDelay.bind(this),
+      options.initialDelayMs || DEFAULT_INITIAL_DELAY_MS
+    );
+  }
+
+  _afterInitialDelay(): void {
+    this._refreshCredits();
+    this._refreshIntervalHandle = setInterval(this._refreshCredits.bind(this), this._refreshIntervalMs);
+    this._initialDelayTimeoutHandle = null;
+  }
+
+  setProcess(process: Process): void {
+    this._uuid = process.uuid || '';
+  }
+
+  isAllowed(operation: string): boolean {
+    if (operation in this._credits) {
+      return this._isAllowed(operation);
+    }
+    // Credits for the operation will be asynchronously fetched
```

Comment:
Just curious, is relying on the interval consistent with the Go and Java throttlers?

Reply:
yup

Reply:
:+1: